File: mobile_apps/flutter_app/lib/services/passkey_service.dart
///
/// TigerWallet Flutter - Passkey Service (WebAuthn)
/// 
/// Complete Passkey/WebAuthn Implementation:
/// - Platform authenticator support
/// - Biometric integration
/// - Secure key storage
/// - Cross-platform compatibility
/// 
/// THIS IS PRODUCTION CODE - NO MOCKS OR SIMULATIONS
/// 

import 'dart:convert';
import 'dart:math' show Random;
import 'dart:typed_data';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/export.dart';

/// Passkey credential types
enum PasskeyAuthenticatorAttachment {
  platform,
  crossPlatform,
}

/// Passkey verification levels
enum PasskeyVerificationLevel {
  none,
  biometric,
  devicePasscode,
}

/// Passkey credential
class PasskeyCredential {
  final String id;
  final String publicKey;
  final String algorithm;
  final String counter;
  final DateTime createdAt;
  final String rpId;

  PasskeyCredential({
    required this.id,
    required this.publicKey,
    required this.algorithm,
    required this.counter,
    required this.createdAt,
    required this.rpId,
  });

  Map<String, dynamic> toJson() => {
    'id': id,
    'publicKey': publicKey,
    'algorithm': algorithm,
    'counter': counter,
    'createdAt': createdAt.toIso8601String(),
    'rpId': rpId,
  };

  factory PasskeyCredential.fromJson(Map<String, dynamic> json) => PasskeyCredential(
    id: json['id'],
    publicKey: json['publicKey'],
    algorithm: json['algorithm'],
    counter: json['counter'],
    createdAt: DateTime.parse(json['createdAt']),
    rpId: json['rpId'],
  );
}

/// Passkey registration options
class PasskeyRegistrationOptions {
  final String rpId;
  final String rpName;
  final String userId;
  final String userName;
  final List<String> pubKeyCredParams;
  final int timeout;
  final List<String> excludeCredentials;

  PasskeyRegistrationOptions({
    required this.rpId,
    required this.rpName,
    required this.userId,
    required this.userName,
    required this.pubKeyCredParams,
    this.timeout = 60000,
    this.excludeCredentials = const [],
  });

  Map<String, dynamic> toJson() => {
    'rp': {
      'id': rpId,
      'name': rpName,
    },
    'user': {
      'id': base64Url.encode(utf8.encode(userId)),
      'name': userName,
      'displayName': userName,
    },
    'pubKeyCredParams': pubKeyCredParams.map((alg) => {
      'type': 'public-key',
      'alg': alg,
    }).toList(),
    'timeout': timeout,
    'excludeCredentials': excludeCredentials.map((cred) => {
      'id': cred,
      'type': 'public-key',
    }).toList(),
  };
}

/// Passkey assertion options
class PasskeyAssertionOptions {
  final String rpId;
  final String challenge;
  final int timeout;
  final List<String> allowedCredentials;

  PasskeyAssertionOptions({
    required this.rpId,
    required this.challenge,
    this.timeout = 60000,
    this.allowedCredentials = const [],
  });

  Map<String, dynamic> toJson() => {
    'rpId': rpId,
    'challenge': challenge,
    'timeout': timeout,
    'allowCredentials': allowedCredentials.map((cred) => {
      'id': cred,
      'type': 'public-key',
    }).toList(),
  };
}

/// Passkey Service - Production Implementation
class PasskeyService {
  static final PasskeyService _instance = PasskeyService._internal();
  factory PasskeyService() => _instance;
  PasskeyService._internal();

  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage(
    aOptions: AndroidOptions(encryptedStorage: true),
    iOptions: IOSOptions(accessibility: KeychainAccessibility.first_unlock),
  );

  static const String _credentialsKey = 'passkey_credentials';
  static const String _enabledKey = 'passkey_enabled';

  final Map<String, PasskeyCredential> _credentials = {};
  bool _isEnabled = false;
  String? _currentUserId;

  /// Initialize the passkey service
  Future<void> initialize() async {
    await _loadCredentials();
    final enabledStr = await _secureStorage.read(key: _enabledKey);
    _isEnabled = enabledStr == 'true';
  }

  /// Check if passkey is available on this device.
  ///
  /// Real platform-authenticator detection requires a WebAuthn platform plugin
  /// (e.g. `flutter_passkey` / `webauthn`). The previous implementation always
  /// returned `true`, falsely advertising passkey support. Until a platform
  /// plugin is wired, this returns `false` (honest).
  Future<bool> isPasskeyAvailable() async {
    return false;
  }

  /// Enable/disable passkey
  Future<bool> setPasskeyEnabled(bool enabled) async {
    _isEnabled = enabled;
    await _secureStorage.write(key: _enabledKey, value: enabled.toString());
    return true;
  }

  /// Check if passkey is enabled
  bool get isPasskeyEnabled => _isEnabled;

  /// Generate registration options for passkey creation
  PasskeyRegistrationOptions generateRegistrationOptions({
    required String userId,
    required String userName,
    String rpId = 'tigerwallet.com',
    String rpName = 'TigerWallet',
  }) {
    _currentUserId = userId;
    
    // Supported algorithm IDs for WebAuthn
    // -7: ES256 (ECDSA with P-256 and SHA-256)
    // -257: RS256 (RSA with PKCS#1 v1.5 and SHA-256)
    return PasskeyRegistrationOptions(
      rpId: rpId,
      rpName: rpName,
      userId: userId,
      userName: userName,
      pubKeyCredParams: ['-7', '-257'],
      timeout: 60000,
      excludeCredentials: _credentials.values
          .where((c) => c.rpId == rpId)
          .map((c) => c.id)
          .toList(),
    );
  }

  /// Generate assertion options for passkey authentication
  PasskeyAssertionOptions generateAssertionOptions({
    required String rpId,
    String? challenge,
  }) {
    final challengeBytes = challenge ?? _generateChallenge();
    return PasskeyAssertionOptions(
      rpId: rpId,
      challenge: base64Url.encode(challengeBytes),
      timeout: 60000,
      allowedCredentials: _credentials.values
          .where((c) => c.rpId == rpId)
          .map((c) => c.id)
          .toList(),
    );
  }

  /// Register a new passkey credential
  Future<bool> registerPasskey({
    required String credentialId,
    required String publicKey,
    required String algorithm,
    required String rpId,
  }) async {
    final credential = PasskeyCredential(
      id: credentialId,
      publicKey: publicKey,
      algorithm: algorithm,
      counter: '0',
      createdAt: DateTime.now(),
      rpId: rpId,
    );

    _credentials[credentialId] = credential;
    await _saveCredentials();
    return true;
  }

  /// Authenticate with passkey
  Future<PasskeyCredential?> authenticateWithPasskey(String credentialId) async {
    final credential = _credentials[credentialId];
    if (credential != null) {
      // Update counter in production
      return credential;
    }
    return null;
  }

  /// Get all credentials for a relying party
  List<PasskeyCredential> getCredentials(String rpId) {
    return _credentials.values.where((c) => c.rpId == rpId).toList();
  }

  /// Remove a credential
  Future<bool> removeCredential(String credentialId) async {
    _credentials.remove(credentialId);
    await _saveCredentials();
    return true;
  }

  /// Remove all credentials for a user
  Future<bool> removeAllCredentials() async {
    _credentials.clear();
    await _saveCredentials();
    return true;
  }

  /// Verify a passkey (WebAuthn) signature.
  ///
  /// WebAuthn assertions sign `authenticatorData || SHA-256(clientDataJSON)`.
  /// For ES256 (alg -7) the signature is a DER-encoded ECDSA signature over
  /// P-256 with SHA-256. This performs REAL verification against the stored
  /// public key (expected as a hex-encoded 64-byte uncompressed P-256 point:
  /// x || y). The previous implementation returned `true` unconditionally —
  /// a full authentication bypass. It now rejects anything it cannot verify.
  Future<bool> verifySignature({
    required String credentialId,
    required Uint8List clientDataHash,
    required Uint8List authenticatorData,
    required Uint8List signature,
  }) async {
    final credential = _credentials[credentialId];
    if (credential == null) return false;

    try {
      final pubHex = credential.publicKey.replaceAll('0x', '');
      final pubBytes = Uint8List.fromList(base64.decode(pubHex));
      // Accept either a raw 64-byte (x||y) point or a 65-byte
      // 0x04||x||y uncompressed point.
      Uint8List xy;
      if (pubBytes.length == 64) {
        xy = pubBytes;
      } else if (pubBytes.length == 65 && pubBytes[0] == 0x04) {
        xy = pubBytes.sublist(1);
      } else {
        return false;
      }
      final q = ECDomainParameters('prime256v1').curve
          .decodePoint(Uint8List.fromList([0x04, ...xy]));
      if (q == null) return false;

      final signedData =
          Uint8List.fromList([...authenticatorData, ...clientDataHash]);
      final verifier = ECDSAVerifier(SHA256Digest());
      verifier.init(false, PublicKeyParameter(ECPublicKey(q)));

      return verifier.verifySignature(signedData, ECSignature.fromDER(signature));
    } catch (e) {
      return false;
    }
  }

  /// Generate a cryptographically random 32-byte challenge.
  ///
  /// Seeds FortunaRandom with 32 bytes from `Random.secure()` (platform
  /// CSPRNG). The previous implementation seeded only with the string
  /// `DateTime.now().millisecondsSinceEpoch` (predictable, low-entropy).
  Uint8List _generateChallenge() {
    final secure = Random.secure();
    final seed =
        Uint8List.fromList(List<int>.generate(32, (_) => secure.nextInt(256)));
    final random = SecureRandom('Fortuna')..seed(KeyParameter(seed));
    return random.nextBytes(32);
  }

  /// Load credentials from secure storage
  Future<void> _loadCredentials() async {
    try {
      final data = await _secureStorage.read(key: _credentialsKey);
      if (data != null) {
        final List<dynamic> jsonList = jsonDecode(data);
        for (final item in jsonList) {
          final credential = PasskeyCredential.fromJson(item);
          _credentials[credential.id] = credential;
        }
      }
    } catch (e) {
      _credentials.clear();
    }
  }

  /// Save credentials to secure storage
  Future<void> _saveCredentials() async {
    final data = jsonEncode(
      _credentials.values.map((c) => c.toJson()).toList(),
    );
    await _secureStorage.write(key: _credentialsKey, value: data);
  }

  /// Get current user ID
  String? get currentUserId => _currentUserId;

  /// Get credential count
  int get credentialCount => _credentials.length;
}

File: browser_extensions/brave_extension/chrome_extension/src/services/master/passkey_service.js
/**
 * Passkey Service - Browser Extension (Chrome)
 * 
 * Complete Passkey/WebAuthn Features:
 * - Platform authenticator support
 * - Biometric integration
 * - Credential management
 * 
 * This service MUST be identical across ALL platforms.
 */

class PasskeyService {
  static instance = null;

  static getInstance() {
    if (!PasskeyService.instance) {
      PasskeyService.instance = new PasskeyService();
      PasskeyService.instance.initialize();
    }
    return PasskeyService.instance;
  }

  constructor() {
    this.credentials = new Map();
    this.isEnabled = false;
    this.rpId = 'tigerwallet.com';
    this.rpName = 'TigerWallet';
    this.currentUserId = null;
  }

  async initialize() {
    await this.loadCredentials();
    const result = await chrome.storage.local.get('passkey_enabled');
    this.isEnabled = result.passkey_enabled || false;
  }

  async loadCredentials() {
    try {
      const result = await chrome.storage.local.get('passkey_credentials');
      if (result.passkey_credentials) {
        result.passkey_credentials.forEach(cred => {
          this.credentials.set(cred.id, cred);
        });
      }
    } catch (e) {
      console.error('Failed to load credentials:', e);
    }
  }

  async saveCredentials() {
    try {
      await chrome.storage.local.set({
        passkey_credentials: Array.from(this.credentials.values())
      });
    } catch (e) {
      console.error('Failed to save credentials:', e);
    }
  }

  // Check if passkey is available
  async isPasskeyAvailable() {
    if (!window.PublicKeyCredential) {
      return false;
    }
    
    try {
      // Check if platform authenticator is available
      const available = await PublicKeyCredential.isUserVerifyingPlatformAuthenticatorAvailable();
      return available || true; // Allow fallback to cross-platform
    } catch (e) {
      return true; // Assume available
    }
  }

  // Enable/disable passkey
  async setPasskeyEnabled(enabled) {
    this.isEnabled = enabled;
    await chrome.storage.local.set({ passkey_enabled: enabled });
    return true;
  }

  isPasskeyEnabled() {
    return this.isEnabled;
  }

  // Generate registration options
  generateRegistrationOptions(userId, userName) {
    this.currentUserId = userId;
    
    // Generate challenge
    const challenge = new Uint8Array(32);
    crypto.getRandomValues(challenge);
    
    return {
      rp: {
        id: this.rpId,
        name: this.rpName
      },
      user: {
        id: this.base64UrlEncode(this.stringToBuffer(userId)),
        name: userName,
        displayName: userName
      },
      pubKeyCredParams: [
        { type: 'public-key', alg: -7 }, // ES256
        { type: 'public-key', alg: -257 } // RS256
      ],
      timeout: 60000,
      excludeCredentials: Array.from(this.credentials.values())
        .filter(c => c.rpId === this.rpId)
        .map(c => ({
          id: this.base64UrlDecode(c.id),
          type: 'public-key'
        }))
    };
  }

  // Generate assertion options
  generateAssertionOptions(rpId, challenge) {
    if (!challenge) {
      challenge = new Uint8Array(32);
      crypto.getRandomValues(challenge);
    }
    
    return {
      rpId: rpId,
      challenge: challenge,
      timeout: 60000,
      allowCredentials: Array.from(this.credentials.values())
        .filter(c => c.rpId === rpId && c.isActive)
        .map(c => ({
          id: this.base64UrlDecode(c.id),
          type: 'public-key'
        }))
    };
  }

  // Register a new credential
  async registerCredential(credentialId, publicKey, algorithm, rpId, userId) {
    const credential = {
      id: credentialId,
      publicKey: publicKey,
      algorithm: algorithm,
      counter: '0',
      rpId: rpId,
      userId: userId,
      createdAt: Date.now(),
      lastUsedAt: Date.now(),
      isActive: true
    };
    
    this.credentials.set(credentialId, credential);
    await this.saveCredentials();
    return true;
  }

  // Authenticate with a credential
  async authenticate(credentialId) {
    const credential = this.credentials.get(credentialId);
    if (!credential) {
      return { success: false, errorMessage: 'Credential not found' };
    }
    
    if (!credential.isActive) {
      return { success: false, errorMessage: 'Credential is inactive' };
    }
    
    // Update counter
    const counter = parseInt(credential.counter) + 1;
    credential.counter = counter.toString();
    credential.lastUsedAt = Date.now();
    
    await this.saveCredentials();
    
    return {
      success: true,
      credentialId: credential.id,
      userId: credential.userId,
      signatureCount: counter
    };
  }

  // Get all credentials for a relying party
  getCredentials(rpId) {
    return Array.from(this.credentials.values())
      .filter(c => c.rpId === rpId && c.isActive);
  }

  // Get credential by ID
  getCredential(credentialId) {
    return this.credentials.get(credentialId);
  }

  // Remove a credential
  async removeCredential(credentialId) {
    const removed = this.credentials.delete(credentialId);
    if (removed) {
      await this.saveCredentials();
    }
    return removed;
  }

  // Remove all credentials
  async removeAllCredentials() {
    this.credentials.clear();
    await this.saveCredentials();
  }

  // Verify signature
  verifySignature(credentialId, clientDataHash, authenticatorData, signature) {
    const credential = this.credentials.get(credentialId);
    if (!credential) return false;
    
    // In production, verify ECDSA signature using stored public key
    // This is a simplified implementation
    if (!signature || signature.length < 64) {
      return false;
    }
    
    return true;
  }

  // Get relying party info
  getRpId() {
    return this.rpId;
  }

  getRpName() {
    return this.rpName;
  }

  getCredentialCount() {
    return this.credentials.size;
  }

  // Helper: String to buffer
  stringToBuffer(str) {
    return new TextEncoder().encode(str);
  }

  // Helper: Buffer to string
  bufferToString(buffer) {
    return new TextDecoder().decode(buffer);
  }

  // Helper: Base64 URL encode
  base64UrlEncode(buffer) {
    const base64 = btoa(String.fromCharCode(...new Uint8Array(buffer)));
    return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  }

  // Helper: Base64 URL decode
  base64UrlDecode(str) {
    const base64 = str.replace(/-/g, '+').replace(/_/g, '/');
    const padded = base64 + '='.repeat((4 - base64.length % 4) % 4);
    const binary = atob(padded);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }
}

// Export for use
if (typeof module !== 'undefined' && module.exports) {
  module.exports = PasskeyService;
}
